### PR TITLE
Tweak GitHub Continuous Integration

### PR DESCRIPTION
**Pull request by atoomic**
This pull request goal is to reduce the amount of resources used on GitHub.

First all smoke workflow are now merged in a single file named `.github/workflows/testsuite.yml`
then we first smoke a single linux flavor and will only smoke the others on success. 

This would avoid to run tests on windows/macos when there is an obvious failure.

This is also introducing the idea that all other workflows should only be smoked when `ENABLE_FULL_TEST_SUITE` secret is set. Avoiding to bump organization quota when test suite is run on forked repositories. (note that secrets cannot be use inside one `if:` condition)

### CODE REVIEW

**Comment by xenu** [comment]
> Avoiding to bump organization quota when test suite is run on forked repositories. (note that secrets cannot be use inside one if: condition)

I was under the impression that github actions resource limits are per repository (for OSS projects) and I can't find any documentation that says otherwise.

**Comment by jkeenan** [comment]
Can we configure this such that if `ENABLE_FULL_TEST_SUITE` is set, we don't get a red `X` next to configs that are not run for that reason?  I see approx 11 such red `X`s above.

**Comment by atoomic** [comment]
This is a WIP and yes the goal is to handle correctly `ENABLE_FULL_TEST_SUITE` when it first run it was not setup in the repo. I will have to keep tweaking it before merging

**Comment by atoomic** [comment]
@xenu we had a discussion with Github and you have quota at organization level... forks can also end up counting in your org... it seems that GitHub is working on providing one UI to let user know in the future where/how the quota are used

File: .github/workflows/testsuite.yml [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+  check_full_test_suite:
+    name: "ENABLE_FULL_TEST_SUITE set?"
+    runs-on: ubuntu-latest
+
+    steps:
+      # --------------------------------------------------------------------------------------------
+      # cannot use secrets earlier as part of 'if:' condition
+      - name: "Runs if ENABLE_FULL_TEST_SUITE is set"
+        run: |
+          if [ -z "${ENABLE_FULL_TEST_SUITE}" ]; then
```
Comment (atoomic) [inline .github/workflows/testsuite.yml]: need to understand why ENABLE_FULL_TEST_SUITE is not set here

**Comment by Leont** [comment]
Per [Lancaster consensus](https://github.com/Perl-Toolchain-Gang/toolchain-site/blob/master/lancaster-consensus.md), toolchain has a concept of extended testing, controlled by the `EXTENDED_TESTING` environmental. I think it would be a good idea to reuse that variable (instead of `ENABLE_FULL_TEST_SUITE`) for consistency throughout our ecosystem.

**Comment by atoomic** [comment]
ok sounds a good idea, will use `EXTENDED_TESTING` instead

File: .github/workflows/testsuite.yml [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+      - name: Run Tests
+        run: |
+          make -j2 test
+
+  #         _         _
+  # __ __ _(_)_ _  __| |_____ __ _____
+  # \ V  V / | ' \/ _` / _ \ V  V (_-<
+  #  \_/\_/|_|_||_\__,_\___/\_/\_//__/
+
+  windows-msvc42:
```
Comment (xenu) [inline .github/workflows/testsuite.yml]: This should be "142", not "42".

Comment (atoomic) [reply .github/workflows/testsuite.yml]: nice catch changed for the next update

File: .github/workflows/testsuite.yml (hunk 1 above)
Comment (xsawyerx) [reply .github/workflows/testsuite.yml]: Was this resolved?

Comment (atoomic) [reply .github/workflows/testsuite.yml]: no this is not solved, this is probably a limitation of secrets, I need to investigate the best way to integrate them in a shell script
I wonder if they are not obfuscated

**Comment by atoomic** [comment]
when testing it using an isolated repository this is working fine...
https://github.com/atoomic/test-workflow/runs/729768257?check_suite_focus=true

I'm not sure what's wrong there...

**Comment by atoomic** [comment]
I see the secret is not set... I wonder if it's using the secrets from the clone repo...
when it's working we can see that`EXTENDED_TESTING: ***`
whereas it's empty otherwise... 
```
env:
   EXTENDED_TESTING: ***
```

**Comment by atoomic** [comment]
I think this is now ready for merge

File: .github/workflows/testsuite.yml [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+    steps:
+      # --------------------------------------------------------------------------------------------
+      # cannot use secrets earlier as part of 'if:' condition
+      - name: "Runs if EXTENDED_TESTING is set"
+        env:
+          EXTENDED_TESTING: ${{ secrets.EXTENDED_TESTING }}
+          PR_TARGET_REPOSITORY: github.event.pull_request.head.repo.full_name == 'Perl/perl5' || github.repository == 'Perl/perl5'
+          CURRENT_REPOSITORY: github.repository == 'Perl/perl5'
+        run: |
+          if [ -z "${EXTENDED_TESTING}" || "${PR_TARGET_REPOSITORY}" eq 'Perl/perl5' || "${CURRENT_REPOSITORY}" eq 'Perl/perl5' ]; then
```
Comment (toddr) [inline .github/workflows/testsuite.yml]: Doesn't this mean that forked repos won't be able to enable extended_testing?

Comment (atoomic) [reply .github/workflows/testsuite.yml]: this is using a `or` logic, which mean by default extended testing are not run on fork unless they set the secret variable EXTENDED_TESTING as recommended by the message

Comment (toddr) [reply .github/workflows/testsuite.yml]: discussed and re-worked.

File: .github/workflows/testsuite.yml (hunk 1 above)
Comment (xsawyerx) [reply .github/workflows/testsuite.yml]: You wrote recently that you think it's ready. Does it mean you resolved this issue?

**Pull request merged**